### PR TITLE
refactor: dependency location

### PR DESCRIPTION
**Pull request by shulaoda**
## Summary
Related to #7622 
> I will start governance again, and some dependencies can be directly addressed using `RealDependencyLocation`, as its location is actually determined. And in some places, we may need to use `SyntheticDependencyLocation` and `DependencyLocation`.
> 
> ```rust
> struct RealDependencyRange {
>   pub end: u32,
>   pub start: u32,
> }
> 
> struct SyntheticDependencyName {
>   pub name: String
> }
> 
> enum DependencyLocation {
>   Real(RealDependencyRange),
>   Synthetic(SyntheticDependencyName),
> }
> ```

Looks good. We might need `index` for `SyntheticDependencyName` as well in some places.

_Originally posted by @h-a-n-a in https://github.com/web-infra-dev/rspack/issues/7622#issuecomment-2300714328_
            

## Checklist

- [ ] Tests updated (or not required).
- [ ] Documentation updated (or not required).

### CODE REVIEW

File: crates/rspack_core/src/dependency/dependency_location.rs [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+        start.line, start.column, end.line, end.column
+      ));
+    }
+
+    None
+  }
+}
+
+#[derive(Debug, Clone)]
+pub struct SyntheticDependencyName {
```
Comment (h-a-n-a) [inline crates/rspack_core/src/dependency/dependency_location.rs:61]: `SyntheticDependencyLocation`?

Comment (h-a-n-a) [reply crates/rspack_core/src/dependency/dependency_location.rs:61]: Let's align the name if there's nothing different between us and webpack.

File: crates/rspack_core/src/dependency/dependency_location.rs [hunk 2]
```diff
@@ -0,0 +1,118 @@
+use std::sync::Arc;
+
+use derivative::Derivative;
+
+#[derive(Derivative)]
+#[derivative(Debug, Clone)]
+pub struct RealDependencyRange {
```
Comment (h-a-n-a) [inline crates/rspack_core/src/dependency/dependency_location.rs:7]: What about `RealDependencyLocation`?

File: crates/rspack_core/src/dependency/dependency_location.rs (hunk 1 above)
Comment (shulaoda) [reply crates/rspack_core/src/dependency/dependency_location.rs:61]: It seems that this structure has been implemented in the past, now let's temporarily use this name. I will resolve the conflict issue in the next PR, which won't be too long.

Comment (h-a-n-a) [reply crates/rspack_core/src/dependency/dependency_location.rs:61]: I see.

**Review by h-a-n-a — APPROVED** [review]
Thanks again!

**Pull request merged**